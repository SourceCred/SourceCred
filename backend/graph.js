// @flow

export type ID = {
  pluginName: string,
  repositoryName: string,
  name: string,
};

export type GraphNode<T> = {
  id: ID,
  edges: ID[],
  payload: T,
};

export type GraphEdge<T> = {
  id: ID,
  sourceId: ID,
  destId: ID,
  weight: number,
  payload: T,
};

export type Graph = {
  nodes: {[stringID: string]: GraphNode<mixed>},
  edges: {[stringID: string]: GraphEdge<mixed>},
};
