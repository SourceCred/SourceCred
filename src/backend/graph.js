// @flow

export type Address = {
  repositoryName: string,
  pluginName: string,
  id: string,
};

export type Node<T> = {
  address: Address,
  inEdges: Address[],
  outEdges: Address[],
  payload: T,
};

export type Edge<T> = {
  address: Address,
  src: Address,
  dst: Address,
  payload: T,
};

export type Graph = {
  nodes: {[stringAddress: string]: Node<mixed>},
  edges: {[stringAddress: string]: Edge<mixed>},
};

export function addressToString(address: Address) {
  if (address.repositoryName.includes("$")) {
    const escaped = JSON.stringify(address.repositoryName);
    throw new Error(`address.repositoryName must not include "\$": ${escaped}`);
  }
  if (address.pluginName.includes("$")) {
    const escaped = JSON.stringify(address.pluginName);
    throw new Error(`address.pluginName must not include "\$": ${escaped}`);
  }
  if (address.id.includes("$")) {
    const escaped = JSON.stringify(address.id);
    throw new Error(`address.id must not include "\$": ${escaped}`);
  }
  return `${address.repositoryName}\$${address.pluginName}\$${address.id}`;
}

export function stringToAddress(string: string) {
  const parts = string.split("$");
  if (parts.length !== 3) {
    const escaped = JSON.stringify(string);
    throw new Error(`Input should have exactly two \$s: ${escaped}`);
  }
  return {
    repositoryName: parts[0],
    pluginName: parts[1],
    id: parts[2],
  };
}
