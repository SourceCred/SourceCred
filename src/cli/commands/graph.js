// @flow

import {Command, flags} from "@oclif/command";
import mkdirp from "mkdirp";
import os from "os";
import path from "path";

import {pluginNames} from "./common";

const execDependencyGraph = require("../../tools/execDependencyGraph").default;

export default class GraphCommand extends Command {
  static description = "create the contribution graph for a repository";

  static args = [
    {
      name: "repo_owner",
      required: true,
      description: "owner of the GitHub repository for which to fetch data",
    },
    {
      name: "repo_name",
      required: true,
      description: "name of the GitHub repository for which to fetch data",
    },
  ];

  static flags = {
    "output-directory": flags.string({
      short: "o",
      description: "directory into which to store graphs",
    }),
    "github-token": flags.string({
      description:
        "a GitHub API token, as generated at " +
        "https://github.com/settings/tokens/new",
      required: true,
      env: "SOURCECRED_GITHUB_TOKEN",
    }),
  };

  async run() {
    const {
      args: {repo_owner: repoOwner, repo_name: repoName},
      flags: {"github-token": token, "output-directory": rawOutputDirectory},
    } = this.parse(GraphCommand);
    const outputDirectory = (() => {
      if (rawOutputDirectory != null) {
        return rawOutputDirectory;
      }
      const outputDirectory = path.join(os.tmpdir(), "sourcecred", repoName);
      this.log("Using output directory: " + outputDirectory);
      return outputDirectory;
    })();
    graph(outputDirectory, repoOwner, repoName, token);
  }
}

function graph(
  outputDirectory: string,
  repoOwner: string,
  repoName: string,
  token: string
) {
  mkdirp.sync(outputDirectory);
  const tasks = makeTasks(outputDirectory, {repoOwner, repoName, token});
  execDependencyGraph(tasks).then(({success}) => {
    process.exitCode = success ? 0 : 1;
  });
}

function makeTasks(outputDirectory, {repoOwner, repoName, token}) {
  const taskId = (id) => `create-${id}`;
  const graphFilename = (id) => path.join(outputDirectory, `graph-${id}.json`);
  const into = "./src/cli/into.sh";
  return [
    ...pluginNames().map((id) => ({
      id: taskId(id),
      cmd: [
        into,
        graphFilename(id),
        "node",
        "./bin/sourcecred.js",
        "plugin-graph",
        "--plugin",
        id,
        repoOwner,
        repoName,
        "--github-token",
        token,
      ],
      deps: [],
    })),
    {
      id: "combine",
      cmd: [
        into,
        path.join(outputDirectory, "graph.json"),
        "node",
        "./bin/sourcecred.js",
        "combine",
        ...pluginNames().map((id) => graphFilename(id)),
      ],
      deps: pluginNames().map((id) => taskId(id)),
    },
  ];
}
