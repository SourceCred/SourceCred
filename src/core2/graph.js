// @flow

import deepEqual from "lodash.isequal";
import type {Address} from "./address";
import {AddressMap} from "./address";
import type {Compatible} from "../util/compat";
import stringify from "json-stable-stringify";

export type Node<NR: NodeReference, NP: NodePayload> = {|
  +ref: NR,
  +payload: NP,
  +address: Address,
|};

export interface NodePayload {
  address(): Address;

  /**
   * Convert this object to its serialized form. This must be a plain
   * old JSON value: i.e., a value `v` such that
   * `JSON.parse(JSON.stringify(v))` is deep-equal to `v`.
   */
  toJSON(): any;
}

export interface NodeReference {
  graph(): Graph;
  address(): Address;
  get(): ?Node<any, any>;

  neighbors(
    options?: NeighborsOptions
  ): Iterator<{|+ref: NodeReference, +edge: Edge<any>|}>;
}

export type Edge<+T> = {|
  +address: Address,
  +src: Address,
  +dst: Address,
  +payload: T,
|};

export type PluginFilter = {|+plugin: string, +type?: string|};
function addressFilterer(options?: PluginFilter) {
  if (options == null) {
    return (_: Address) => true;
  }
  const {plugin, type} = options;
  if (plugin == null) {
    throw new Error("PluginFilters must filter by plugin");
  }
  return ({owner}: Address) =>
    owner.plugin === plugin && (type == null || owner.type === type);
}
export type NeighborsOptions = {|
  +node?: PluginFilter,
  +edge?: PluginFilter,
  +direction?: "IN" | "OUT" | "ANY",
|};

export interface PluginHandler<NR: NodeReference, NP: NodePayload> {
  /**
   * Enrich a base reference with plugin-/domain-specific properties.
   */
  createReference(baseReference: NodeReference): NR;

  /**
   * Deserialize a JSON payload, which is guaranteed to be the
   * serialization of a previous `NP`.
   */
  createPayload(json: any): NP;

  /**
   * Provide the name of the plugin.
   * Should return a constant string.
   */
  pluginName(): string;
}

export type Plugins = $ReadOnlyArray<PluginHandler<any, any>>;

type MaybeNode = {|+address: Address, +node: Node<any, any> | void|};
type Integer = number;

type IndexedEdge = {|
  +address: Address,
  +srcIndex: Integer,
  +dstIndex: Integer,
  +payload: any,
|};

export class Graph {
  _plugins: Plugins;
  _pluginMap: PluginMap;

  // Invariant: sizes of `_nodeIndices`, `_nodes`, `_outEdges`, and
  // `_inEdges` are all equal.
  _nodes: MaybeNode[];
  _nodeIndices: AddressMap<{|+address: Address, +index: Integer|}>;

  // If `idx` is the index of a node `v`, then `_outEdges[idx]` is the
  // list of `e.address` for all edges `e` whose source is `v`.
  // Likewise, `_inEdges[idx]` has the addresses of all in-edges to `v`.
  _outEdges: Address[][];
  _inEdges: Address[][];
  _edges: AddressMap<IndexedEdge>;

  constructor(plugins: Plugins) {
    this._plugins = plugins.slice();
    this._pluginMap = createPluginMap(this._plugins);
    this._nodes = [];
    this._nodeIndices = new AddressMap();
    this._outEdges = [];
    this._inEdges = [];
    this._edges = new AddressMap();
  }

  ref(address: Address): NodeReference {
    if (address == null) {
      throw new Error(`address is ${String(address)}`);
    }
    // If node has an index and is still present, return the existing ref
    const indexDatum = this._nodeIndices.get(address);
    if (indexDatum != null) {
      const node = this._nodes[indexDatum.index].node;
      if (node != null) {
        return node.ref;
      }
    }
    // Otherwise, create a "dummy ref" that isn't backed by a node.
    const handler = findHandler(this._pluginMap, address.owner.plugin);
    return handler.createReference(new InternalReference(this, address));
  }

  node(address: Address): ?Node<any, any> {
    return this.ref(address).get();
  }

  /**
   * Get nodes in the graph, in unspecified order.
   *
   * If filter is provided, it will return only nodes with the requested plugin name
   * (and, optionally, type).
   */
  *nodes(options?: PluginFilter): Iterator<Node<any, any>> {
    const filter = addressFilterer(options);
    for (const maybeNode of this._nodes) {
      const node = maybeNode.node;
      if (node == null) {
        continue;
      }
      if (!filter(node.address)) {
        continue;
      }
      yield node;
    }
  }

  edge(address: Address): ?Edge<any> {
    const indexedEdge = this._edges.get(address);
    if (!indexedEdge) {
      return undefined;
    }
    return this._upgradeIndexedEdge(indexedEdge);
  }

  _upgradeIndexedEdge(indexedEdge: IndexedEdge): Edge<any> {
    return {
      address: indexedEdge.address,
      src: this._nodes[indexedEdge.srcIndex].address,
      dst: this._nodes[indexedEdge.dstIndex].address,
      payload: indexedEdge.payload,
    };
  }

  /**
   * Gets edges in the graph, in unspecified order.
   *
   * If filter is provided, it will return only edges with the requested type.
   */
  *edges(options?: PluginFilter): Iterator<Edge<any>> {
    let edges = this._edges
      .getAll()
      .map((indexedEdge) => this._upgradeIndexedEdge(indexedEdge));
    const filter = addressFilterer(options);
    for (const edge of edges) {
      if (filter(edge.address)) {
        yield edge;
      }
    }
  }

  _addNodeAddress(address: Address): Integer {
    const indexDatum = this._nodeIndices.get(address);
    if (indexDatum != null) {
      return indexDatum.index;
    } else {
      const index = this._nodes.length;
      this._nodeIndices.add({address, index});
      this._nodes.push({address, node: undefined});
      this._outEdges.push([]);
      this._inEdges.push([]);
      return index;
    }
  }

  addNode(payload: NodePayload): this {
    if (payload == null) {
      throw new Error(`payload is ${String(payload)}`);
    }
    const address = payload.address();
    const index = this._addNodeAddress(address);
    const maybeNode = this._nodes[index];
    if (maybeNode.node !== undefined) {
      if (deepEqual(maybeNode.node.payload, payload)) {
        return this;
      } else {
        throw new Error(
          `node at address ${JSON.stringify(
            address
          )} exists with distinct contents`
        );
      }
    }
    const handler = findHandler(this._pluginMap, address.owner.plugin);
    const ref = handler.createReference(new InternalReference(this, address));
    const node = {ref, payload, address};
    this._nodes[index] = {address, node};
    return this;
  }

  removeNode(address: Address): this {
    const indexDatum = this._nodeIndices.get(address);
    if (indexDatum != null) {
      this._nodes[indexDatum.index] = {address, node: undefined};
    }
    return this;
  }

  addEdge(edge: Edge<any>): this {
    if (edge == null) {
      throw new Error(`edge is ${String(edge)}`);
    }
    if (edge.address == null) {
      throw new Error(`address is ${String(edge.address)}`);
    }
    if (edge.src == null) {
      throw new Error(`src is ${String(edge.src)}`);
    }
    if (edge.dst == null) {
      throw new Error(`dst is ${String(edge.dst)}`);
    }
    const srcIndex = this._addNodeAddress(edge.src);
    const dstIndex = this._addNodeAddress(edge.dst);
    const indexedEdge = {
      address: edge.address,
      srcIndex,
      dstIndex,
      payload: edge.payload,
    };
    return this._addIndexedEdge(indexedEdge);
  }

  _addIndexedEdge(indexedEdge: IndexedEdge): this {
    const existingIndexedEdge = this._edges.get(indexedEdge.address);
    if (existingIndexedEdge !== undefined) {
      if (deepEqual(existingIndexedEdge, indexedEdge)) {
        return this;
      } else {
        throw new Error(
          `edge at address ${JSON.stringify(
            indexedEdge.address
          )} exists with distinct contents`
        );
      }
    }
    this._edges.add(indexedEdge);
    this._outEdges[indexedEdge.srcIndex].push(indexedEdge.address);
    this._inEdges[indexedEdge.dstIndex].push(indexedEdge.address);
    return this;
  }

  removeEdge(address: Address): this {
    // TODO(perf): This is linear in the degree of the endpoints of the
    // edge. Consider storing in non-list form.
    const indexedEdge = this._edges.get(address);
    if (indexedEdge) {
      this._edges.remove(address);
      [
        this._outEdges[indexedEdge.srcIndex],
        this._inEdges[indexedEdge.dstIndex],
      ].forEach((edges) => {
        const index = edges.findIndex((ea) => deepEqual(ea, address));
        if (index !== -1) {
          edges.splice(index, 1);
        }
      });
    }
    return this;
  }

  /**
   * Merge a collection of graphs. If multiple graphs have a node with a
   * particular address, then the nodes must all have identical payload,
   * and a single copy of this node will be included in the result
   * graph; if not all nodes are identical, then an error is thrown.
   * Likewise, all edges at a particular address must have identical
   * source, destination, and payload.
   *
   * The existing graph objects are not modified.
   */
  static mergeConservative(
    plugins: Plugins,
    graphs: $ReadOnlyArray<Graph>
  ): Graph {
    const result = new Graph(plugins);
    graphs.forEach((graph) => {
      for (const node of graph.nodes()) {
        result.addNode(node.payload);
      }
      for (const edge of graph.edges()) {
        result.addEdge(edge);
      }
    });
    return result;
  }

  /**
   * Check the equality of two graphs. This verifies that the node and edge
   * contents are identical; it does not check which plugin handlers are
   * registered.
   */
  equals(that: Graph): boolean {
    const theseNodes = Array.from(this.nodes());
    const thoseNodes = Array.from(that.nodes());
    if (theseNodes.length !== thoseNodes.length) {
      return false;
    }

    const theseEdges = Array.from(this.edges());
    const thoseEdges = Array.from(that.edges());
    if (theseEdges.length !== thoseEdges.length) {
      return false;
    }

    for (const node of theseNodes) {
      if (!deepEqual(node, that.node(node.address))) {
        return false;
      }
    }
    for (const edge of theseEdges) {
      if (!deepEqual(edge, that.edge(edge.address))) {
        return false;
      }
    }
    return true;
  }

  copy(): Graph {
    return Graph.mergeConservative(this.plugins(), [this]);
  }

  plugins(): Plugins {
    return this._plugins.slice();
  }

  static fromJSON(plugins: Plugins, json: Compatible<GraphJSON>): Graph {
    const _ = {plugins, json};
    throw new Error("Graphv2 is not yet implemented");
  }

  toJSON(): Compatible<GraphJSON> {
    throw new Error("Graphv2 is not yet implemented");
  }
}

export type GraphJSON = any;

type PluginMap = {[pluginName: string]: PluginHandler<any, any>};
function createPluginMap(plugins: Plugins): PluginMap {
  const pluginMap = {};
  plugins.forEach((p) => {
    const name = p.pluginName();
    if (pluginMap[name] != null) {
      throw new Error(`Duplicate plugin handler for "${name}"`);
    }
    pluginMap[name] = p;
  });
  return pluginMap;
}
function findHandler(pluginMap: PluginMap, pluginName: string) {
  if (pluginMap[pluginName] == null) {
    throw new Error(`No plugin handler for "${pluginName}"`);
  }
  return pluginMap[pluginName];
}

const DELEGATE_NODE_REFERENCE_BASE = Symbol("base");
function getBase(dnr: DelegateNodeReference): NodeReference {
  // Flow doesn't know about Symbols, so we use this function to
  // localize the `any`-casts as much as possible.
  return (dnr: any)[DELEGATE_NODE_REFERENCE_BASE];
}

export class DelegateNodeReference implements NodeReference {
  constructor(base: NodeReference) {
    (this: any)[DELEGATE_NODE_REFERENCE_BASE] = base;
  }
  graph() {
    return getBase(this).graph();
  }
  address() {
    return getBase(this).address();
  }
  get() {
    return getBase(this).get();
  }
  neighbors(options?: NeighborsOptions) {
    return getBase(this).neighbors(options);
  }
}

class InternalReference implements NodeReference {
  _graph: Graph;
  _address: Address;

  constructor(graph: Graph, address: Address) {
    this._graph = graph;
    this._address = address;
  }

  graph(): Graph {
    return this._graph;
  }
  address(): Address {
    return this._address;
  }
  get(): ?Node<any, any> {
    const indexDatum = this._graph._nodeIndices.get(this._address);
    if (indexDatum == null) {
      return undefined;
    }
    return this._graph._nodes[indexDatum.index].node;
  }

  *neighbors(
    options?: NeighborsOptions
  ): Iterator<{|+ref: NodeReference, +edge: Edge<any>|}> {
    const indexDatum = this._graph._nodeIndices.get(this._address);
    if (indexDatum == null) {
      return;
    }
    const nodeIndex = indexDatum.index;

    const direction = (options && options.direction) || "ANY";
    const edgeFilter =
      options == null ? (_) => true : addressFilterer(options.edge);
    const nodeFilter =
      options == null ? (_) => true : addressFilterer(options.node);

    const graph = this._graph;
    const adjacencies = [];
    if (direction === "ANY" || direction === "IN") {
      adjacencies.push({list: graph._inEdges[nodeIndex], direction: "IN"});
    }
    if (direction === "ANY" || direction === "OUT") {
      adjacencies.push({list: graph._outEdges[nodeIndex], direction: "OUT"});
    }

    for (const adjacency of adjacencies) {
      for (const edgeAddress of adjacency.list) {
        const indexedEdge = graph._edges.get(edgeAddress);
        if (indexedEdge == null) {
          throw new Error(
            `Edge at address ${stringify(edgeAddress)} does not exist`
          );
        }
        if (direction === "ANY" && adjacency.direction === "IN") {
          if (indexedEdge.srcIndex === indexedEdge.dstIndex) {
            continue;
          }
        }
        const edge = graph._upgradeIndexedEdge(indexedEdge);
        const ref = graph.ref(
          adjacency.direction === "IN" ? edge.src : edge.dst
        );
        if (edgeFilter(edge.address) && nodeFilter(ref.address())) {
          yield {edge, ref};
        }
      }
    }
  }
}
