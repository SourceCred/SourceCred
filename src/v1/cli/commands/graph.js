// @flow

import {Command, flags} from "@oclif/command";
import mkdirp from "mkdirp";
import path from "path";

import {
  pluginNames,
  sourcecredDirectoryFlag,
  nodeMaxOldSpaceSizeFlag,
} from "../common";

const execDependencyGraph = require("../../../tools/execDependencyGraph")
  .default;

export default class GraphCommand extends Command {
  static description = `\
create the contribution graph for a repository

Create the contribution graph for a repository. This creates a
contribution graph for each individual plugin, and then combines the
individual graphs into one larger graph. The graphs are stored as JSON
files under SOURCECRED_DIRECTORY/graphs/REPO_OWNER/REPO_NAME.
`.trim();

  static args = [
    {
      name: "repo_owner",
      required: true,
      description: "owner of the GitHub repository for which to fetch data",
    },
    {
      name: "repo_name",
      required: true,
      description: "name of the GitHub repository for which to fetch data",
    },
  ];

  static flags = {
    "sourcecred-directory": sourcecredDirectoryFlag(),
    "max-old-space-size": nodeMaxOldSpaceSizeFlag(),
    "github-token": flags.string({
      description:
        "a GitHub API token, as generated at " +
        "https://github.com/settings/tokens/new",
      required: true,
      env: "SOURCECRED_GITHUB_TOKEN",
    }),
  };

  async run() {
    const {
      args: {repo_owner: repoOwner, repo_name: repoName},
      flags: {
        "github-token": token,
        "sourcecred-directory": sourcecredDirectory,
        "max-old-space-size": maxOldSpaceSize,
      },
    } = this.parse(GraphCommand);
    graph(sourcecredDirectory, repoOwner, repoName, token, maxOldSpaceSize);
  }
}

function graph(
  sourcecredDirectory: string,
  repoOwner: string,
  repoName: string,
  token: string,
  maxOldSpaceSize: number
) {
  const graphDirectory = path.join(
    sourcecredDirectory,
    "graphs",
    repoOwner,
    repoName
  );
  console.log("Storing graphs into: " + graphDirectory);
  mkdirp.sync(graphDirectory);
  const tasks = makeTasks(graphDirectory, {
    repoOwner,
    repoName,
    token,
    maxOldSpaceSize,
  });
  execDependencyGraph(tasks, {taskPassLabel: "DONE"}).then(({success}) => {
    process.exitCode = success ? 0 : 1;
  });
}

function makeTasks(
  graphDirectory,
  {repoOwner, repoName, token, maxOldSpaceSize}
) {
  const taskId = (id) => `create-${id}`;
  const graphFilename = (id) => path.join(graphDirectory, `graph-${id}.json`);
  const into = "./src/v1/cli/into.sh";
  return [
    ...pluginNames().map((id) => ({
      id: taskId(id),
      cmd: [
        into,
        graphFilename(id),
        "node",
        `--max_old_space_size=${maxOldSpaceSize}`,
        "./bin/sourcecred.js",
        "plugin-graph",
        "--plugin",
        id,
        repoOwner,
        repoName,
        "--github-token",
        token,
      ],
      deps: [],
    })),
    {
      id: "combine",
      cmd: [
        into,
        path.join(graphDirectory, "graph.json"),
        "node",
        `--max_old_space_size=${maxOldSpaceSize}`,
        "./bin/sourcecred.js",
        "combine",
        ...pluginNames().map((id) => graphFilename(id)),
      ],
      deps: pluginNames().map((id) => taskId(id)),
    },
  ];
}
