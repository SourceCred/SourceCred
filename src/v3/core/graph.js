// @flow

import type {_NodeAddress, _EdgeAddress} from "./address";
import {_Address} from "./address";

export const Address = _Address;
export type NodeAddress = _NodeAddress;
export type EdgeAddress = _EdgeAddress;
export type Edge = {|
  +address: EdgeAddress,
  +src: NodeAddress,
  +dst: NodeAddress,
|};

export type Neighbor = {|+node: NodeAddress, +edge: Edge|};
export opaque type Direction = Symbol;
export const IN: Direction = Symbol("IN");
export const OUT: Direction = Symbol("OUT");
export type NeighborsOptions = {|
  +direction: ?Direction,
  +nodePrefix: ?NodeAddress,
  +edgePrefix: ?EdgeAddress,
|};

export opaque type GraphJSON = any; // TODO

export class Graph {
  _nodes: Set<NodeAddress>;
  // If `e` is an Edge in the graph, then:
  // * _edges.get(e.address) `deepEquals` e
  // * _inEdges.get(e.dst) `contains` e
  // * _outEdges.get(e.src) `contains` e
  _edges: Map<EdgeAddress, Edge>;
  _inEdges: Map<NodeAddress, Edge[]>;
  _outEdges: Map<NodeAddress, Edge[]>;

  constructor(): void {
    this._nodes = new Set();
    this._edges = new Map();
    this._inEdges = new Map();
    this._outEdges = new Map();
  }

  addNode(a: NodeAddress): this {
    const _ = a;
    throw new Error("addNode");
  }

  removeNode(a: NodeAddress): this {
    const _ = a;
    throw new Error("removeNode");
  }

  hasNode(a: NodeAddress): boolean {
    const _ = a;
    throw new Error("hasNode");
  }

  nodes(): Iterator<NodeAddress> {
    throw new Error("nodes");
  }

  addEdge({src, dst, address}: Edge): this {
    const _ = {src, dst, address};
    throw new Error("addEdge");
  }

  removeEdge(a: EdgeAddress): this {
    const _ = a;
    throw new Error("removeEdge");
  }

  hasEdge(address: EdgeAddress): boolean {
    const _ = address;
    throw new Error("hasEdge");
  }

  edge(address: EdgeAddress): ?Edge {
    const _ = address;
    throw new Error("edge");
  }

  edges(): Iterator<Edge> {
    throw new Error("edges");
  }

  neighbors(node: NodeAddress, options?: NeighborsOptions): Iterator<Neighbor> {
    const _ = {node, options};
    throw new Error("neighbors");
  }

  copy(): Graph {
    throw new Error("copy");
  }

  toJSON(): GraphJSON {
    throw new Error("toJSON");
  }

  static fromJSON(json: GraphJSON): Graph {
    const _ = json;
    throw new Error("fromJSON");
  }

  static merge(graphs: Iterable<Graph>): Graph {
    const _ = graphs;
    throw new Error("merge");
  }
}
