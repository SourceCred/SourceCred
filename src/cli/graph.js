// @flow

import fs from "fs-extra";
import sortBy from "lodash.sortby";
import stringify from "json-stable-stringify";
import {join as pathJoin} from "path";
import {loadFileWithDefault} from "../util/disk";

import {Ledger} from "../ledger/ledger";
import * as NullUtil from "../util/null";
import {LoggingTaskReporter} from "../util/taskReporter";
import {type ReferenceDetector} from "../core/references/referenceDetector";
import {CascadingReferenceDetector} from "../core/references/cascadingReferenceDetector";
import type {Command} from "./command";
import dedent from "../util/dedent";
import {type InstanceConfig} from "../api/instanceConfig";
import {
  makePluginDir,
  loadInstanceConfig,
  pluginDirectoryContext,
} from "./common";
import {toJSON as weightedGraphToJSON} from "../core/weightedGraph";
import * as pluginId from "../api/pluginId";
import {ensureIdentityExists} from "../ledger/identityProposal";

function die(std, message) {
  std.err("fatal: " + message);
  return 1;
}

const graphCommand: Command = async (args, std) => {
  const taskReporter = new LoggingTaskReporter();
  taskReporter.start("graph");
  const baseDir = process.cwd();
  const config: InstanceConfig = await loadInstanceConfig(baseDir);

  const ledgerPath = pathJoin(baseDir, "data", "ledger.json");
  const ledger = Ledger.parse(
    await loadFileWithDefault(ledgerPath, () => new Ledger().serialize())
  );

  let pluginsToLoad = [];
  if (args.length === 0) {
    pluginsToLoad = config.bundledPlugins.keys();
  } else {
    for (const arg of args) {
      const id = pluginId.fromString(arg);
      if (config.bundledPlugins.has(id)) {
        pluginsToLoad.push(id);
      } else {
        return die(
          std,
          `can't find plugin ${id}; remember to use fully scoped name, as in sourcecred/github`
        );
      }
    }
  }

  const graphOutputPrefix = ["output", "graphs"];

  const rd = await buildReferenceDetector(
    baseDir,
    config,
    taskReporter,
    ledger
  );
  for (const name of pluginsToLoad) {
    const plugin = NullUtil.get(config.bundledPlugins.get(name));
    const task = `${name}: generating graph`;
    taskReporter.start(task);
    const dirContext = pluginDirectoryContext(baseDir, name);
    const graph = await plugin.graph(dirContext, rd, taskReporter);
    const serializedGraph = stringify(weightedGraphToJSON(graph));
    const outputDir = makePluginDir(baseDir, graphOutputPrefix, name);
    const outputPath = pathJoin(outputDir, "graph.json");
    await fs.writeFile(outputPath, serializedGraph);

    const identities = await plugin.identities(dirContext, taskReporter);
    for (const identityProposal of identities) {
      ensureIdentityExists(ledger, identityProposal);
    }
    taskReporter.finish(task);
  }
  await fs.writeFile(ledgerPath, ledger.serialize());
  taskReporter.finish("graph");
  return 0;
};

async function buildReferenceDetector(baseDir, config, taskReporter, ledger) {
  taskReporter.start("reference detector");
  const rds = [];
  for (const [name, plugin] of sortBy(
    [...config.bundledPlugins],
    ([k, _]) => k
  )) {
    const dirContext = pluginDirectoryContext(baseDir, name);
    const task = `reference detector for ${name}`;
    taskReporter.start(task);
    const rd = await plugin.referenceDetector(dirContext, taskReporter);
    rds.push(rd);
    taskReporter.finish(task);
  }
  taskReporter.finish("reference detector");
  rds.push(_hackyIdentityNameReferenceDetector(ledger));
  return new CascadingReferenceDetector(rds);
}

// Hack to support old-school (deprecated) "initiatives files":
// We need to be able to parse references to usernames, e.g. "@yalor", so
// we need a reference detector that will match these to identities in the
// Ledger. This isn't a robust addressing scheme, since identities are re-nameable;
// in v2 the initiatives plugin will be re-written to use identity node addresses instead.
// This hack can be safely deleted once we no longer support initiatives files that refer
// to identities by their names instead of their IDs.
export function _hackyIdentityNameReferenceDetector(
  ledger: Ledger
): ReferenceDetector {
  const usernameToAddress = new Map(
    ledger.accounts().map((a) => [a.identity.name, a.identity.address])
  );
  function addressFromUrl(potentialUsername: string) {
    const prepped = potentialUsername.replace("@", "").toLowerCase();
    return usernameToAddress.get(prepped);
  }
  return {addressFromUrl};
}

export const graphHelp: Command = async (args, std) => {
  std.out(
    dedent`\
      usage: sourcecred graph

      Generate a graph from cached plugin data

      Either 'sourcecred load' must immediately precede this command or
      a cache directory must exist for all plugins specified in sourcecred.json
      `.trimRight()
  );
  return 0;
};

export default graphCommand;
