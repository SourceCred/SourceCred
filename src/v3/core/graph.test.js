// @flow

import {Address, Graph} from "./graph";
import type {NodeAddress, EdgeAddress} from "./graph";

describe("core/graph", () => {
  describe("Address re-exports", () => {
    it("include the Address psuedo-module", () => {
      expect(Address).toEqual(expect.anything());
    });
    it("include distinct NodeAddress and EdgeAddress types", () => {
      const nodeAddress: NodeAddress = Address.nodeAddress();
      const edgeAddress: EdgeAddress = Address.edgeAddress();
      // $ExpectFlowError
      const badNodeAddress: NodeAddress = edgeAddress;
      // $ExpectFlowError
      const badEdgeAddress: EdgeAddress = nodeAddress;
      const _ = {badNodeAddress, badEdgeAddress};
    });
  });

  describe("Graph class", () => {
    it("can be constructed", () => {
      const x = new Graph();
      // Verify that `x` is not of type `any`
      // $ExpectFlowError
      expect(() => x.measureSpectacularity()).toThrow();
    });
  });
});
