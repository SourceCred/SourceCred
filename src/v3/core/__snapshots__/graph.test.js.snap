// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/graph toJSON / fromJSON toJSON matches snapshot 1`] = `
Array [
  Object {
    "type": "sourcecred/graph",
    "version": "0.4.0",
  },
  Object {
    "edges": Array [
      Object {
        "address": Array [
          "edge",
        ],
        "dstIndex": 0,
        "srcIndex": 1,
      },
      Object {
        "address": Array [
          "edge",
          "2",
        ],
        "dstIndex": 1,
        "srcIndex": 0,
      },
    ],
    "nodes": Array [
      Array [
        "dst",
      ],
      Array [
        "src",
      ],
    ],
  },
]
`;
